File: 572-subtree-of-another-tree/subtree-of-another-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} subRoot
 * @return {boolean}
 */
var isSubtree = function(root, subRoot) {
    const sameTree = (s, t) => {
        if (!s && !t) {
            return true
        } 
        if (s && t && s.val === t.val) {
            let left = sameTree(s.left, t.left)
            let right = sameTree(s.right, t.right)
            return (left && right)
        }
    }

    if (!subRoot) {
        return true
    } else if (!root && subRoot) {
        return false
    }
    if (sameTree(root, subRoot)) {
        return true
    }
    let left = isSubtree(root.left, subRoot)
    let right = isSubtree(root.right, subRoot)
    return left || right
};

File: 572-subtree-of-another-tree/subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:

        def isSameTree(p, q):
            if not p and not q:
                return True
            
            if p is None and q is not None:
                return False
            
            if q is None and p is not None:
                return False

            if p.val == q.val:    
                left = isSameTree(p.left, q.left)
                right = isSameTree(p.right, q.right)
                return left and right
            else:
                return False
        
        def checkTrees(p, q):
            if p is None:
                return False
            
            if isSameTree(p,q):
                return True
            
            return checkTrees(p.left, q) or checkTrees(p.right, q)

        return checkTrees(root,subRoot)